foundation-models/openai/src/test/java/com/sap/ai/sdk/foundationmodels/openai/OpenAiClientTest.java

```diff
@@ -128,122 +159,309 @@ void testErrorHandling() {
+
+      final var request = new OpenAiEmbeddingParameters().setInput("Hello World");
+      final var result = client.embedding(request);
+
+      assertThat(result).isNotNull();
+      assertThat(result.getModel()).isEqualTo("ada");
+      assertThat(result.getObject()).isEqualTo("list");
+
+      assertThat(result.getUsage()).isNotNull();
+      assertThat(result.getUsage().getCompletionTokens()).isNull();
+      assertThat(result.getUsage().getPromptTokens()).isEqualTo(2);
+      assertThat(result.getUsage().getTotalTokens()).isEqualTo(2);
+
+      assertThat(result.getData()).isNotNull().hasSize(1);
+      var embeddingData = result.getData().get(0);
+      assertThat(embeddingData).isNotNull();
+      assertThat(embeddingData.getObject()).isEqualTo("embedding");
+      assertThat(embeddingData.getIndex()).isEqualTo(0);
+      assertThat(embeddingData.getEmbedding())
+          .isNotNull()
+          .isNotEmpty()
+          .containsExactly(-0.0000000070958645d, 2.123e-300d, -0.0069813123d, -3.385849e-05d)
+          // ensure double precision
+          .hasToString("[-7.0958645E-9, 2.123E-300, -0.0069813123, -3.385849E-5]");
+
+      verify(
+          postRequestedFor(urlPathEqualTo("/embeddings"))
+              .withRequestBody(
+                  equalToJson("""
+                      {"input":["Hello World"]}""")));
+    }
+  }
+
+  @Test
+  void streamChatCompletionErrorHandling() throws IOException {
+    try (var inputStream =
+        spy(
+            Objects.requireNonNull(
+                getClass()
+                    .getClassLoader()
+                    .getResourceAsStream("streamChatCompletionError.txt")))) {
+
+      final var httpClient = mock(HttpClient.class);
+      ApacheHttpClient5Accessor.setHttpClientFactory(destination -> httpClient);
+
+      // Create a mock response
+      final var mockResponse = new BasicClassicHttpResponse(200, "OK");
+      final var inputStreamEntity = new InputStreamEntity(inputStream, ContentType.TEXT_PLAIN);
+      mockResponse.setEntity(inputStreamEntity);
+      mockResponse.setHeader("Content-Type", "text/event-stream");
+
+      // Configure the HttpClient mock to return the mock response
+      doReturn(mockResponse).when(httpClient).executeOpen(any(), any(), any());
+
+      final var request =
+          new OpenAiChatCompletionParameters()
+              .setMessages(
+                  List.of(
+                      new OpenAiChatUserMessage()
+                          .addText(
+                              "Can you give me the first 100 numbers of the Fibonacci sequence?")));
+
+      try (Stream<OpenAiChatCompletionDelta> stream = client.streamChatCompletion(request)) {
+        assertThatThrownBy(() -> stream.forEach(System.out::println))
+            .isInstanceOf(OpenAiClientException.class)
+            .hasMessage(
+                "Failed to parse response from OpenAI model and error message: 'exceeded token rate limit'");
```

Comment:
do you know if this can actually happen mid-stream?

Reply:
I experienced it once, I wrote it down [here](https://github.com/SAP/ai-sdk-java/pull/25/commits/69ae7eb220f19dd973330d3b6ce0e47c569fd3b7#diff-8f8ce52bc428755b7c06bb07061be2021bf64675307b39101ba85b3549f35019R44-R50)

Reply:
Thanks, I'll attach this to a related BLI I am creating.